### ISSUE TITLE
Resources not assigned correctly when using lambda functions + execution groups

### ISSUE THREAD

**Issue by jonaszierer** [snakemake/snakemake#2045]
**Snakemake version**
7.18.2

**Describe the bug**
I use lambda functions for resource assignment (eg. `resources: mem_mb = lambda w, attempt: 4000*attempt`). This works great, however when using execution groups the lambda function is not evaluated correctly, and the memory is not increased with increasing attempts.

**Minimal example**

```
rule a:
    output: "a_{number}"
    resources: mem_mb = lambda w, attempt: attempt*1000
    group: "mygroup"
    shell: '''
    if [ {resources.mem_mb} == 3000 ]; then
        touch -m {output}
    fi
    '''

rule b:
    input: rules.a.output
    output: "b_{number}"
    group: "mygroup"
    shell: '''
    touch -m {output}
    '''
```

Now one can run `snakemake --profile <your cluster profile> b_5 --cores 1 --restart-times 3`. Snakemake should fail with MissingOutputException in the first 2 attempts (when mem_mb is 1000 and 2000, respectively, a_5 will not be created). It should eventually restart the job with 3000MB and succeed. This works without the group assignments (or when running locally) but it stops working when using groups. When using groups, snakemake does start the job 3 times but the memory remains the same

**Comment by pvandyken** [comment]
I'm able to replicate. It looks like the issue arises due to some caching that's not being appropriately reset in the group context. I'll work up a patch as soon as I'm able.

**Comment by pvandyken** [comment]
My proposed fix: #2052. Feel free to give it a try to see if it resolves in your environment.
